### DISCUSSION TITLE
How to call a trpc mutation from outsize a component

### DISCUSSION THREAD

**Discussion by TheBestMoshe** [trpc/trpc#1351] [Q&A]
I am setting up a refresh and access token for my app. The app retrieves the refresh token when the user signs in. The tokens are stored in Redux.

The access token expires after 5 minutes. Before every request, I would like to check if we have a valid access token, and if not, I would like to fetch the access token before making the request.

I would also like to extract all the logic into a separate function so that I can just call `getAccessToken` and it can retrieve it.

How can I access the trpc client from within a function which is not part of a component?

Here is my client creation function:

```
...
    trpc.createClient({
      url: `${url}/api/trpc`,

      async headers() {
        let accessToken = store.getState().user.accessToken;
        const refreshToken = store.getState().user.refreshToken

        if (!accessToken && refreshToken) { // TODO check if accessToken and refresh token is valid
          // I would like to call a mutation to fetch the access token
        }

        const authHeaders = accessToken ? `Bearer ${accessToken}` : undefined

        return {
          'Authorization': authHeaders
        };
      },
      transformer,
    }),
...
```

As a stopgap, I am going to try using `fetch` to create the request myself. Is there any way to do this with trpc natively?

**Comment by KATT** [comment]
You could store it as a global variable or on the `window` but the nicest thing to do would be to use it thorough the context (`trpc.useContext().client`)

**Reply by TheBestMoshe** [reply]
> but the nicest thing to do would be to use it thorough the context (trpc.useContext().client)

Isn't `useContext` a React hook? I want to call outside of a component.

I tried it outside a component and I'm getting an error:

```
Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
```

>You could store it as a global variable

What should be stored in a global variable? The client created by `trpc.createClient`? I'm getting a type of `any` for the client. Is there any way to get the correct type? Then I could store it in a global, and access it outside a component.

![Screen Shot 2021-12-17 at 11 14 08 AM](https://user-images.githubusercontent.com/34072688/146574647-46b9d232-daa2-45c0-996c-d7d326799b94.png)

**Reply by KATT** [reply]
- When you call `trpc.createClient()` you can store that outside too. 
- Unsure why you're getting `any`, I can't debug that without access to the project.
- On holiday now (#1346 ) and will shut down completely any day now. 
- Feel free to ask in the help channel on trpc.io/discord - a bunch of people that are pretty active there.

**Reply by TheBestMoshe** [reply]
My project is based on the [zART stack template](https://github.com/KATT/zart). I just checked, and `trpc.createClient` in the template also is of type `any`.

I will open an issue for it, and check out the Discord.

Enjoy your vacation!! tRPC is really amazing! Thanks!

**Reply by jetaggart** [reply]
@TheBestMoshe did you ever figure this out? I have a similar issue, I have the client stored as a global, but outside of react context I'm not sure how to call queries, and I get no types to guide me.

**Reply by TenchuBlossom** [reply]
Hey guys. I ran into this problem too on trpc V10. I found you can create an additional vanilla `createTRPCProxyClient` and simply use that to call your trpc procedures from outside of a react component. Using the vanilla client also gives you strong types too. For all other calls to your backend that are inside react components you can use the trpc created from `createTRPCReact`. Its a bit of a hack and should be considered a temporary work-around, Hope this helps :)

**Links** 
https://trpc.io/docs/v10/vanilla
https://trpc.io/docs/v10/react

**Example** 
```
// create vanilla trpc client
const trpcNonReact = createTRPCProxyClient<AppRouter>({
  links: [
    httpBatchLinkVanilla({
      url: "http://localhost:8080/trpc",
    }),
  ],
});

// call vanilla trpc client from outside of react
const callTRPCOutsideReact = async () => {
  const result = await trpcNonReact.login.refresh.mutate({body: {refreshToken: 'refresh-token-goes-here'}})
}
```

**Reply by jetaggart** [reply]
@TenchuBlossom I think the problem is the cache with react-query will not be in sync with this other instance. Ideally they're bound together so the cache is not out of sync.

**Reply by TheBestMoshe** [reply]
> @TheBestMoshe did you ever figure this out?

@bit-void I never did find a solution to this

**Reply by TenchuBlossom** [reply]
> @TenchuBlossom I think the problem is the cache with react-query will not be in sync with this other instance. Ideally they're bound together so the cache is not out of sync.

Yeah thats correct, the solution i suggested is a hack for sure. but if you only need to make a few isolated calls outside of a component,  to get a new refresh token and then add that to a zustand store for example, the above solution will work for now. But we definitely need a full working solution for trpc/react outside of components. The alternative is to simply port your code to the vanilla client (createTRPCProxyClient) and just use that, you will then have to manually wrap this up in your own implementation of react query. But at least you will get some typesaftey, the ability to call outside of components  and a single cache.

**Reply by jetaggart** [reply]
For sure, I think the real issue lies here: https://github.com/trpc/trpc/discussions/2926. We can pull a client off the react instance but it's incorrectly typed.

**Reply by TenchuBlossom** [reply]
Yeah incorrect types are deffo the issue. Thanks for sharing. hope we can get a fix asap
